### PR TITLE
Issue #1217 Enable update code resilient towards exact location of minishift binary

### PR DESCRIPTION
**Pull request by praveenkumar**
Todo
-------

- Need to update unit test to accommodate all those changes.

### CODE REVIEW

File: pkg/minishift/update/update.go [hunk 1]
```diff
@@ -264,12 +267,34 @@ func extractBinary(downloadedArchivePath, archiveDir string) (string, error) {
...
+func minishiftBinaryPath(extractPath, binaryName string) (string, error) {
+	var binaryPath string
+	var err error
+	fileList := []string{}
+	err = filepath.Walk(extractPath, func(path string, f os.FileInfo, err error) error {
+		fileList = append(fileList, path)
+		return nil
+	})
+	for _, fileName := range fileList {
+		if strings.Contains(fileName, binaryName) {
```
Comment (budhrg) [inline pkg/minishift/update/update.go]: I doubt we can use this login in `filepath.Walk...` invocation above and avoid this loop. 

What are the content of `fileList` ?

Comment (hferentschik) [inline pkg/minishift/update/update.go]: I am not quite sure what @budhrg means with _login_, but afaict you walk the tree and append everything to a slice. Then you iterate the slice to see whether you got a match. Why not do everything in one go in the anonymous function passed to file path.Walk? Just check each entry whether it is a match. If so record the entry. After file path.Walk you just check whether you found something. 

Up to you what you want to do.

Comment (budhrg) [inline pkg/minishift/update/update.go]: Typo here :) It is `lines`. Corrected above.

**Comment by praveenkumar** [comment]
>  would expect that we try at least with a current archive structure, the one we are aiming for (one more intermediate directory) and one with non existent binary.

I agree so currently test case contain first 2 and I will add non-existent binary tarball which should fail to provide binary path.

> Checking in all these binaries can blows up the repo quite quickly.

@hferentschik these tarballs are dummy if you check the size it's just same as a empty file.

**Comment by hferentschik** [comment]
> these tarballs are dummy if you check the size it's just same as a empty file.

ahh, perfect. That's fine then. I did not check the size.

**Comment by praveenkumar** [comment]
@hferentschik, @budhrg  Please do the review, I am done adding respective test cases.

**Comment by hferentschik** [comment]
ok

**Comment by hferentschik** [comment]
@praveenkumar Seems some of your tests are not working on Windows.

File: pkg/util/archive/archive.go [hunk 2]
```diff
@@ -58,27 +58,50 @@ func Untar(tarball, targetDir string) error {
 
 	for {
 		header, err := tarReader.Next()
-		if err == io.EOF {
-			break
-		} else if err != nil {
+		switch {
```
Comment (hferentschik) [inline pkg/util/archive/archive.go]: what drives these changes? A bug in the untar?

File: pkg/minishift/update/update_test.go [hunk 3]
```diff
@@ -84,15 +89,73 @@ func TestExtractBinary(t *testing.T) {
...
+		osName = "windows"
+	}
+	archiveName := fmt.Sprintf("minishift-%s-%s-%s.%s", "1.2.0", osName, "amd64", extName)
 	testArchivePath := filepath.Join(testDataDir, archiveName)
 	tmpArchivePath := filepath.Join(testDir, archiveName)
 	// Copy test archive to tmpDir for testing purpose
 	copyArchive(t, testArchivePath, tmpArchivePath)
-	extractedBinaryPath, err := extractBinary(tmpArchivePath, testDir)
+	fileName, err := extractBinary(tmpArchivePath, testDir)
+	t.Log(fileName)
```
Comment (hferentschik) [inline pkg/minishift/update/update_test.go]: t.Log?

File: pkg/minishift/update/update_test.go [hunk 4]
```diff
@@ -84,15 +89,73 @@ func TestExtractBinary(t *testing.T) {
...
+	if !strings.HasSuffix(fileName, "minishift") {
+		t.Fatalf("Expected minishift binary path, Got : %s\n", fileName)
```
Comment (hferentschik) [inline pkg/minishift/update/update_test.go]: On Windows we have minishift.exe

File: pkg/minishift/update/update_test.go [hunk 5]
```diff
@@ -84,15 +89,73 @@ func TestExtractBinary(t *testing.T) {
 		extName = "zip"
 		osName = "windows"
 	}
+	archiveName := fmt.Sprintf("minishift-%s-%s-%s.%s", "1.0.0", osName, "amd64", extName)
+	testArchivePath := filepath.Join(testDataDir, archiveName)
+	tmpArchivePath := filepath.Join(testDir, archiveName)
+	// Copy test archive to tmpDir for testing purpose
+	copyArchive(t, testArchivePath, tmpArchivePath)
+	fileName, err := extractBinary(tmpArchivePath, testDir)
+	t.Log(fileName)
```
Comment (hferentschik) [inline pkg/minishift/update/update_test.go]: I am not sure whether we should add more noise to the test output. Generally we should aim to have as little was possible get mixed into the test output.

File: pkg/minishift/update/update_test.go [hunk 6]
```diff
@@ -84,15 +89,73 @@ func TestExtractBinary(t *testing.T) {
...
+	if !strings.HasSuffix(fileName, "minishift") {
```
Comment (hferentschik) [inline pkg/minishift/update/update_test.go]: minishift.exe on Windows

File: pkg/util/archive/archive.go (hunk 2 above)
Comment (praveenkumar) [inline pkg/util/archive/archive.go]: @hferentschik yes the way we implemented it first time only thinking about our tarball structure which had files in first level but when I tried same implementation it break when files are underneath of a directory structure.

Comment (hferentschik) [inline pkg/util/archive/archive.go]: got you. makes sense

**Comment by hferentschik** [comment]
Rebased and merged, thanks.

**Pull request closed without merging**